test/functional/ui/messages_spec.lua

```diff
@@ -22,8 +22,129 @@ describe('ui/ext_messages', function()
+        kind = "confirm"
+      }, {
+        content = { { "1" } },
+        kind = "echo"
+      }, {
+        content = { { "Press ENTER or type command to continue", 4 } },
+        kind = "return_prompt"
+    } }}
+    feed('<cr><cr>')
+
+    -- kind=confirm_sub
+    feed(':%s/i/X/gc<cr>')
+    screen:expect{grid=[[
+      l{7:i}ne 1                   |
+      l{8:i}ne ^2                   |
+      {1:~                        }|
+      {1:~                        }|
+      {1:~                        }|
+    ]], attr_ids={
+      [1] = {bold = true, foreground = Screen.colors.Blue1},
+      [2] = {foreground = Screen.colors.Grey100, background = Screen.colors.Red},
+      [3] = {bold = true},
+      [4] = {bold = true, foreground = Screen.colors.SeaGreen4},
+      [5] = {foreground = Screen.colors.Blue1},
+      [6] = {bold = true, reverse = true},
+      [7] = {reverse = true},
+      [8] = {background = Screen.colors.Yellow},
+    }, messages={ {
+        content = { { "replace with X (y/n/a/q/l/^E/^Y)?", 4 } },
+        kind = "confirm_sub"
+      } }}
+    feed('nq')
+
+    -- kind=wmsg (editing readonly file)
+    command('write Xtest')
+    command('set readonly nohls')
+    feed('G$x')
+    screen:expect{grid=[[
+        line 1                   |
+        {IGNORE}|
+        {1:~                        }|
+        {1:~                        }|
+        {1:~                        }|
+      ]], attr_ids={
+      [1] = {bold = true, foreground = Screen.colors.Blue1},
+      [7] = {foreground = Screen.colors.Red},
+      }, messages={ {
+        content = { { "W10: Warning: Changing a readonly file", 7 } },
+        kind = "wmsg"
+      }
+    }}
+
+    -- kind=wmsg ('wrapscan' after search reaches EOF)
+    feed('uG$/i<cr>')
+    screen:expect{grid=[[
+      l^ine 1                   |
+      line 2                   |
+      {1:~                        }|
+      {1:~                        }|
+      {1:~                        }|
+    ]], attr_ids={
+      [1] = {bold = true, foreground = Screen.colors.Blue1},
+      [2] = {foreground = Screen.colors.Grey100, background = Screen.colors.Red},
+      [3] = {bold = true},
+      [4] = {bold = true, foreground = Screen.colors.SeaGreen4},
+      [5] = {foreground = Screen.colors.Blue1},
+      [6] = {bold = true, reverse = true},
+      [7] = {foreground = Screen.colors.Red},
+    }, messages={ {
+        content = { { "search hit BOTTOM, continuing at TOP", 7 } },
+        kind = "wmsg"
+      } }}
+
+    -- kind=emsg after :throw
+    feed(':throw "foo"<cr>')
+    screen:expect{grid=[[
+      l^ine 1                   |
+      line 2                   |
+      {1:~                        }|
+      {1:~                        }|
+      {1:~                        }|
+    ]], messages={ {
+        content = { { "Error detected while processing :", 2 } },
+        kind = "emsg"
+      }, {
+        content = { { "E605: Exception not caught: foo", 2 } },
+        kind = ""
```

Comment:
The first message is marked as `"emsg"` but this one is blank `""` :( Why doesn't the `do_cmdline` change apply?

Ignoring this for now.